### DISCUSSION TITLE
Can not send confirmation email on selfhosted instance

### DISCUSSION THREAD

**Discussion by Loigzorn** [simple-login/app#1193] [Self-hosting question]
Hello everyone and thanks a lot to the team of simple-login for implementing such an amazing app.

I followed the steps of the README to self host my own instance of simple-login.
Finally I was at the point to be able to access the webgui at app.mydomain.com.
But it seems that there is still a small mismatch in the config as I am unable to send a confirmation email to sign up with an email address.

## Prerequisites
- [x] I have searched open and closed issues to make sure that the bug has not yet been reported.

## Bug report

**Describe the bug**
Unable to send a confirmation as descripted above. 

**Expected behavior**
Receiving the confirmation email to complete registration and start using SimpleLogin.

**Environment**
- OS: Ubuntu 22.04 on a V-Server with 2 GB of RAM and sufficient CPU for the setup
- Browser: Chrome, Brave (Chromium-based)
- Version: postgres:12.1, simplelogin/app:3.4.0 (as recommended in the README)

**Additional context**
I host the simpleLogin service behind a reverse proxy, therefore I use the docker image of the [NginxProxyManager](https://nginxproxymanager.com/setup/).

The descripted issue is already topic of several bug reports, so far none of the solutions has worked for me.

As soon as I try to sign up a email address like `myemail@gmail.com` I will run in a worker timeout as seen in the log below:
```bash
>>> URL: https://app.mydomain.com
Paddle param not set
Upload files to local dir
>>> init logging <<<
2022-07-26 13:37:49,235 - SL - DEBUG - "/code/app/utils.py:11" - <module>() -  - load words file: /code/local_data/words_alpha.txt
2022-07-26 13:37:49,954 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 91.57.55.20 GET /auth/resend_activation ImmutableMultiDict([]) 200
2022-07-26 14:18:08,394 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 91.57.55.20 GET /auth/register ImmutableMultiDict([]) 200
2022-07-26 14:18:21,940 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 91.57.55.20 POST /auth/register ImmutableMultiDict([]) 200
2022-07-26 14:18:47,797 - SL - DEBUG - "/code/app/auth/views/register.py:77" - register() -  - create user myemail@gmail.com
2022-07-26 14:18:48,360 - SL - DEBUG - "/code/app/email_utils.py:250" - send_email() -  - send email to myemail@gmail.com, subject Just one more step to join SimpleLogin
[2022-07-26 14:19:03 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:47)
[2022-07-26 14:19:03 +0000] [47] [INFO] Worker exiting (pid: 47)
2022-07-26 14:19:04,023 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 91.57.55.20 GET /favicon.ico ImmutableMultiDict([]) 302
[2022-07-26 14:19:04 +0000] [91] [INFO] Booting worker with pid: 91
```

the same error occures when trying to resend a activation email:
```bash
>>> URL: https://app.mydomain.com
Paddle param not set
Upload files to local dir
>>> init logging <<<
2022-07-26 13:35:59,819 - SL - DEBUG - "/code/app/utils.py:11" - <module>() -  - load words file: /code/local_data/words_alpha.txt
2022-07-26 13:36:03,200 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 91.57.55.20 GET /auth/register ImmutableMultiDict([]) 200
2022-07-26 13:36:05,168 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 91.57.55.20 GET /auth/login ImmutableMultiDict([]) 200
2022-07-26 13:36:17,722 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 91.57.55.20 POST /auth/login ImmutableMultiDict([]) 200
2022-07-26 13:37:22,474 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 91.57.55.20 GET /auth/resend_activation ImmutableMultiDict([]) 200
2022-07-26 13:37:32,378 - SL - DEBUG - "/code/app/auth/views/resend_activation.py:32" - resend_activation() -  - user <User 10  myemail@gmail.com> is not activated
2022-07-26 13:37:32,410 - SL - DEBUG - "/code/app/email_utils.py:250" - send_email() -  - send email to myemail@gmail.com, subject Just one more step to join SimpleLogin
[2022-07-26 13:37:47 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:73)
[2022-07-26 13:37:47 +0000] [73] [INFO] Worker exiting (pid: 73)
2022-07-26 13:37:47,971 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 91.57.55.20 GET /favicon.ico ImmutableMultiDict([]) 302
[2022-07-26 13:37:48 +0000] [82] [INFO] Booting worker with pid: 82
>>> URL: https://app.mydomain.com
Paddle param not set
Upload files to local dir
>>> init logging <<<
2022-07-26 13:37:49,235 - SL - DEBUG - "/code/app/utils.py:11" - <module>() -  - load words file: /code/local_data/words_alpha.txt
2022-07-26 13:37:49,954 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 91.57.55.20 GET /auth/resend_activation ImmutableMultiDict([]) 200
```

This behaviour was already subject of [Issus 735](https://github.com/simple-login/app/issues/735), so I followed the steps pointed out and double checked the config files at `/etc/postfix/pgsql-relay-domains.cf` and `/etc/postfix/pgsql-transport-maps.cf`. They gave back the expected results as descripted in the [B1 section](https://github.com/simple-login/app/blob/cbcae31288ff68c036e052fb8d04255a6e276ba5/docs/troubleshooting.md#b1) of the troubleshooting docs.
Further it is pointed out to be related to the hardware setup, but I don't think that an issue in my case.
At last I have set the Cloudflare DNS in the `simplelogin.env` and restarted the container:
```env
NAMESERVERS="1.1.1.1", "1.0.0.1"
```

Going on to the next issue: [Issue#319](https://github.com/simple-login/app/issues/319#issuecomment-725962022)
the log output of postfix at `/var/log/mail.log` gave me no reason to doubt the issue is related to the postfix setup:
```log
Jul 26 14:36:30 ubuntu postfix[25926]: Postfix is running with backwards-compatible default settings
Jul 26 14:36:30 ubuntu postfix[25926]: See http://www.postfix.org/COMPATIBILITY_README.html for details
Jul 26 14:36:30 ubuntu postfix[25926]: To disable backwards compatibility use "postconf compatibility_level=3.6" and "postfix reload"
Jul 26 14:36:30 ubuntu postfix/postfix-script[25932]: refreshing the Postfix mail system
Jul 26 14:36:30 ubuntu postfix/master[18221]: reload -- version 3.6.4, configuration /etc/postfix
```

`sudo docker exec -it sl-app python shell.py` in combination with `print(User.query.first())` returned the expected 

```bash
<User 13 myemail@gmail.com>
```

The command
```python
send_email(
                    "myemail@gmail.com",
                    "Test Subject",
                    "Test Body"
                )
```
on the other hand timed out with this response:
```log
/code/app/email_utils.py in send_email(to_email, subject, plaintext, html, unsubscribe_link, unsubscribe_via_email)
    254         smtp.starttls()
    255     else:
--> 256         smtp = SMTP(POSTFIX_SERVER, POSTFIX_PORT or 25)
    257
    258     msg = MIMEMultipart("alternative")

/usr/local/lib/python3.7/smtplib.py in __init__(self, host, port, local_hostname, timeout, source_address)
    249
    250         if host:
--> 251             (code, msg) = self.connect(host, port)
    252             if code != 220:
    253                 self.close()

/usr/local/lib/python3.7/smtplib.py in connect(self, host, port, source_address)
    334         if self.debuglevel > 0:
    335             self._print_debug('connect:', (host, port))
--> 336         self.sock = self._get_socket(host, port, self.timeout)
    337         self.file = None
    338         (code, msg) = self.getreply()

/usr/local/lib/python3.7/smtplib.py in _get_socket(self, host, port, timeout)
    305             self._print_debug('connect: to', (host, port), self.source_address)
    306         return socket.create_connection((host, port), timeout,
--> 307                                         self.source_address)
    308
    309     def connect(self, host='localhost', port=0, source_address=None):

/usr/local/lib/python3.7/socket.py in create_connection(address, timeout, source_address)
    726     if err is not None:
    727         try:
--> 728             raise err
    729         finally:
    730             # Break explicitly a reference cycle

/usr/local/lib/python3.7/socket.py in create_connection(address, timeout, source_address)
    714             if source_address:
    715                 sock.bind(source_address)
--> 716             sock.connect(sa)
    717             # Break explicitly a reference cycle
    718             err = None

TimeoutError: [Errno 110] Connection timed out

```

So my guess is that `sl-app` or `sl-email` is unable to connect to postfix.
Then I tried `docker exec -it sl-app telnet 10.0.0.1 25` in both the container `sl-app` and `sl-email` which both had the response:
```log
Trying 10.0.0.1...
Connected to 10.0.0.1.
Escape character is '^]'.
220 app.mydomain.com ESMTP Postfix (Ubuntu)
```
This gave me confidence they can both connect to postfix (on the outside of the `sl-network`).

The best clue I currently have is a misconfiguration at the docker `sl-network` as the sl-email container does not show any logs of an email going through:
```log
>>> URL: https://app.mydomain.com
Paddle param not set
Upload files to local dir
>>> init logging <<<
2022-07-26 14:36:29,807 - SL - DEBUG - "/code/app/utils.py:11" - <module>() -  - load words file: /code/local_data/words_alpha.txt
2022-07-26 14:36:31,937 - SL - INFO - "email_handler.py:1672" - <module>() -  - Listen for port 20381
2022-07-26 14:36:31,942 - SL - DEBUG - "email_handler.py:1653" - main() -  - Start mail controller 0.0.0.0 20381
```

The command `swaks --to myemail@gmail.com` fails as well as the command with a protonmail address:
```log
=== Trying mail.protonmail.ch:25...
=== Connected to mail.protonmail.ch.
<-  220-mailin025.protonmail.ch ESMTP Postfix
<-  220 mailin025.protonmail.ch ESMTP Postfix
 -> EHLO ubuntu
<-  250-mailin025.protonmail.ch
<-  250-PIPELINING
<-  250-SIZE 36480000
<-  250-STARTTLS
<-  250-ENHANCEDSTATUSCODES
<-  250-8BITMIME
<-  250 CHUNKING
 -> MAIL FROM:<root@ubuntu>
<-  250 2.1.0 Ok
 -> RCPT TO:<otheremail@pm.me>
<** 504 5.5.2 <root@ubuntu>: Sender address rejected: need fully-qualified address
 -> QUIT
<-  221 2.0.0 Bye
=== Connection closed with remote host.
```
But works flawless as soon as I provide the sender address with: `swaks --to myemail@protonmail.com --from root@mydomain.com`

`sl-network` setup:

`sl-db` at 10.0.0.2
`sl-app` at 10.0.0.4
`sl-email` at 10.0.0.5
`sl-job-runner` at 10.0.0.6

Can you point me at the peace of config to take a look at? As I seem to be unable to figure it out by myself.
I would love to self host this fantastic service and hope to hear back from one of you soon.
Your help is appreciated.

**Comment by nguyenkims** [comment]
@Loigzorn I'd suggest to redo the self hosting part as it seems to have issues in Postfix (`swaks` doesn't send outgoing emails) and docker network connection to Postfix. The docker network setup is also quite error prone.

**Comment by strausmann** [comment]
Unfortunately I have the same problem. No mails are going out from SimpleLogin. From the Docker containers, the postfix on the host is accessible without problems.

**Reply by nguyenkims** [reply]
Have you tried the latest version which is https://github.com/simple-login/app/tree/4.6.2-beta ?

**Reply by strausmann** [reply]
Yes, i use the 4.6.2-beta...

**Reply by nguyenkims** [reply]
As I understand:
- you can emails from Postfix to an external mailbox
- you can't however send emails from SimpleLogin container: the container doesn't seem to be able to connect to Postfix.

In this case, this seems to be a docker network issue and I'd suggest to remove the `sl-network` and retry the setup.

**Reply by strausmann** [reply]
I can send Mails from postfix to external Servers. And I can send Mails from the Mail Container to the docker host IP without any Problems.
I found the faq article for this issue. All the Tests was successful.

The sl-network is two times recreated.
I can send Mails to the postfix from external and SimpleLogin processing the Incoming Mails correctly. But outgoing does not works.

**Reply by nguyenkims** [reply]
Can you check if the `POSTFIX_SERVER` set in your SimpleLogin config file is correctly set inside the container? You can check it by connecting to the container python (for ex `docker exec -it sl-email python`), then `import config; print(config.POSTFIX_SERVER)`

**Comment by wisefella** [comment]
Same problem:  Unable to send email.  Looks like SMTP/25 outbound is timing out.  All Outbound Firewall traffic is enabled
`swaks --to jackma@gmail.com` results in:
        sl@simplelogin:~$ swaks --to jackma@gmail.com   //also tried swaks --to jackma@hotmail.com
        === Trying gmail-smtp-in.l.google.com:25...
        *** Error connecting to gmail-smtp-in.l.google.com:25:
        ***     IO::Socket::INET6: connect: timeout
Do we not need to configure SMTP Relay host in main.cf?

Thanks
-wf
